function-sidecar/cmd/function-sidecar.go

```diff
@@ -108,31 +111,75 @@ func main() {
 		defer consumer.Close()
 	}
 
-	dispatcher, err := createDispatcher(protocol)
-	if err != nil {
-		panic(err)
-	}
-	switch d := dispatcher.(type) {
-	case io.Closer:
-		log.Print("Requesting close()")
-		defer d.Close()
-	}
-
-	carrier.Run(consumer, producer, dispatcher, output)
-
 	// trap SIGINT and SIGTERM to trigger a shutdown.
 	signals := make(chan os.Signal, 1)
 	signal.Notify(signals, os.Interrupt, syscall.SIGTERM)
-	<-signals
-	log.Println(("Shutting Down..."))
+
+LOOP:
+	for {
+
+		select {
+		case <-signals:
+			log.Println("Shutting Down...")
+			break LOOP
+		default:
+		}
+
+		log.Print("Creating dispatcher")
+		dispatcher, err := createDispatcher(protocol)
+		if err != nil {
+			if !backoff() {
+				panic(err)
+			} else {
+				continue LOOP
+			}
+		}
+		switch d := dispatcher.(type) {
+		case io.Closer:
+			log.Print("Requesting close()")
+			defer d.Close()
+		}
+
+		carrier.Run(consumer, producer, dispatcher, output)
+
+		select {
+		case <-signals:
+			log.Println("Shutting Down...")
+			break LOOP
+		case <-dispatcher.Closed():
+			log.Println("End of Stream...")
+			if !backoff() {
+				break LOOP
+			}
+		}
+
+	}
+
+}
+
+func backoff() bool {
+	if exitOnComplete {
+		return false
+	} else {
+		// Back off a bit to give the invoker time to come back
+		// (if we support windowing or polling this logic will be more complex)
+		time.Sleep(1000 * time.Millisecond)
+		return true
```

Comment:
Shouldn't the back-off eventually give up? Otherwise we might be stuck in this cycle indefinitely.